packages/pigeon/lib/dart_generator.dart

```diff
@@ -736,6 +1824,249 @@ PlatformException _createConnectionError(String channelName) {
+      final Iterable<String> argExpressions =
+          parameters.mapIndexed((int index, NamedType type) {
+        final String name = _getParameterName(index, type);
+        if (type.type.isEnum) {
+          return '$name${type.type.isNullable ? '?' : ''}.index';
+        } else {
+          return name;
+        }
+      });
+      sendArgument = '<Object?>[${argExpressions.join(', ')}]';
+    }
+
+    indent
+        .writeln("const String ${_varNamePrefix}channelName = '$channelName';");
+    indent.writeScoped(
+        'final BasicMessageChannel<Object?> ${_varNamePrefix}channel = BasicMessageChannel<Object?>(',
+        ');', () {
+      indent.writeln('${_varNamePrefix}channelName,');
+      indent.writeln('$_pigeonChannelCodec,');
+      indent.writeln('binaryMessenger: ${_varNamePrefix}binaryMessenger,');
+    });
+    final String returnTypeName = _makeGenericTypeArguments(returnType);
+    final String genericCastCall = _makeGenericCastCall(returnType);
+    const String accessor = '${_varNamePrefix}replyList[0]';
+    // Avoid warnings from pointlessly casting to `Object?`.
+    final String nullablyTypedAccessor = returnTypeName == 'Object'
+        ? accessor
+        : '($accessor as $returnTypeName?)';
+    final String nullHandler =
+        returnType.isNullable ? (genericCastCall.isEmpty ? '' : '?') : '!';
+    String returnStatement = 'return';
+    if (returnType.isEnum) {
+      if (returnType.isNullable) {
+        returnStatement =
+            '$returnStatement ($accessor as int?) == null ? null : $returnTypeName.values[$accessor! as int]';
+      } else {
+        returnStatement =
+            '$returnStatement $returnTypeName.values[$accessor! as int]';
+      }
+    } else if (!returnType.isVoid) {
+      returnStatement =
+          '$returnStatement $nullablyTypedAccessor$nullHandler$genericCastCall';
+    }
+    returnStatement = '$returnStatement;';
+
+    indent.format('''
+final List<Object?>? ${_varNamePrefix}replyList =
+\t\tawait ${_varNamePrefix}channel.send($sendArgument) as List<Object?>?;
+if (${_varNamePrefix}replyList == null) {
+\tthrow _createConnectionError(${_varNamePrefix}channelName);
+} else if (${_varNamePrefix}replyList.length > 1) {
+\tthrow PlatformException(
+\t\tcode: ${_varNamePrefix}replyList[0]! as String,
+\t\tmessage: ${_varNamePrefix}replyList[1] as String?,
+\t\tdetails: ${_varNamePrefix}replyList[2],
+\t);''');
+    // On iOS we can return nil from functions to accommodate error
+    // handling.  Returning a nil value and not returning an error is an
+    // exception.
+    if (!returnType.isNullable && !returnType.isVoid) {
+      indent.format('''
+} else if (${_varNamePrefix}replyList[0] == null) {
+\tthrow PlatformException(
+\t\tcode: 'null-error',
+\t\tmessage: 'Host platform returned null value for non-null return value.',
+\t);''');
+    }
+    indent.format('''
+} else {
+\t$returnStatement
+}''');
+  }
+
+  void _writeFlutterMethodMessageHandler(
```

Comment:
Some of these incidental refactors could ideally be done in a prequel PR that would be trivial to review.